File: <100/040/040.py
import sys
sys.setrecursionlimit(10**8)


def get_d(n: int,
          q: list[int],
          l: int,
          ans: int
          ) -> int:
    if not q:
        return ans
    
    sn = str(n)
    sl = len(sn)
    bk = q[-1]
    if bk <= l + sl:
        q.pop()
        for i in range(sl):
            if l + i + 1 == bk:
                d = int(sn[i])
                return get_d(n+1, q, l+sl, d*ans)
    
    return get_d(n+1, q, l+sl, ans)


def main() -> int:
    q = [10**j for j in range(6, -1, -1)]
    return get_d(1, q, 0, 1)


if __name__ == '__main__':
    answer = main()
    print(answer)

File: <100/040/040.cpp
#include<iostream>
#include<queue>
using namespace std;


int get_d(int n, queue<int> &q, int l = 0, int ans=1){
  if(q.empty()){
    return ans;
  }

  string sn = to_string(n);
  int sl = sn.size();
  int fr = q.front();
  if(fr <= l + sl){
    q.pop();
    for(int i = 0; i < sl; ++i){
      if(l + i + 1 == fr){
        int d = sn[i] - '0';
        return get_d(n+1, q, l+sl, d*ans);
      }
    }
  }

  return get_d(n+1, q, l+sl, ans);
}


int main(){
  queue<int> q;
  q.push(1); q.push(10); q.push(100); q.push(1000); q.push(10000); q.push(100000); q.push(1000000);
  cout << get_d(1, q) << endl;
  return 0;
}
